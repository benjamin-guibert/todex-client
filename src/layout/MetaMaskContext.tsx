import { createContext, useRef, useState } from 'react'
import { Web3Provider } from '@ethersproject/providers'
import { MetaMaskHandler, initialize, connect, getAccount, getBalance } from 'libraries/metamask'

export enum InitializationStatus {
  None = 0,
  Initializing = 1,
  Initialized = 2,
  NotInstalled = 3,
  Error = 4,
}

export interface MetaMaskContextValue {
  initializationStatus: InitializationStatus
  account: string | undefined
  balance: string | undefined
  initialize: () => Promise<Web3Provider | undefined>
  connect: () => Promise<string | undefined>
  getAccount: () => Promise<string | undefined>
  getBalance: (account: string) => Promise<string | undefined>
}

export const useMetaMaskContext = (): MetaMaskContextValue => {
  const metaMaskHandlerRef = useRef<MetaMaskHandler>()
  const [initializationStatus, setInitializationStatus] = useState<InitializationStatus>(InitializationStatus.None)
  const [account, setAccount] = useState<string | undefined>()
  const [balance, setBalance] = useState<string | undefined>()

  const initializeValue = async () => {
    try {
      metaMaskHandlerRef.current = await initialize()
      if (!metaMaskHandlerRef.current) {
        setInitializationStatus(InitializationStatus.NotInstalled)
      }

      setInitializationStatus(InitializationStatus.Initialized)

      getAccountValue()
    } catch {
      setInitializationStatus(InitializationStatus.Error)
    }

    return metaMaskHandlerRef.current?.provider
  }

  const connectValue = async () => {
    if (!metaMaskHandlerRef.current) {
      return undefined
    }

    const account = await connect(metaMaskHandlerRef.current)
    setAccount(account)

    return account
  }

  const getAccountValue = async () => {
    if (!metaMaskHandlerRef.current) {
      return undefined
    }

    const account = await getAccount(metaMaskHandlerRef.current)
    setAccount(account)

    return account
  }

  const getBalanceValue = async (account: string) => {
    if (!metaMaskHandlerRef.current) {
      return undefined
    }

    const balance = (await getBalance(metaMaskHandlerRef.current, account))?.toString()

    setBalance(balance)

    return balance
  }

  return {
    initializationStatus,
    account,
    balance,
    initialize: initializeValue,
    connect: connectValue,
    getAccount: getAccountValue,
    getBalance: getBalanceValue,
  }
}

export const MetaMaskContext = createContext<MetaMaskContextValue>({
  initializationStatus: InitializationStatus.None,
  account: undefined,
  balance: undefined,
  initialize: async () => undefined,
  connect: async () => undefined,
  getAccount: async () => undefined,
  getBalance: async () => undefined,
})
