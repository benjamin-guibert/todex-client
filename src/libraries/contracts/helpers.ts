import { BigNumber } from '@ethersproject/bignumber'

export const BURN_ADDRESS = '0x0000000000000000000000000000000000000000'
export const DECIMALS = 18

export const getDateFromUnixTimestamp = (timestamp: BigNumber) => {
  return new Date(timestamp.mul(1000).toNumber())
}
