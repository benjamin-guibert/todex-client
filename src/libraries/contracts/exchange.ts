import { BigNumber, constants, Contract, providers } from 'ethers'
import { Web3Provider } from '@ethersproject/providers'
import orderBy from 'lodash/orderBy'
import Exchange from 'contracts/abis/Exchange.json'
import { Exchange as ExchangeType } from 'contracts/Exchange'
import { Token } from 'contracts/Token'
import { DECIMALS, getDateFromUnixTimestamp } from './helpers'
import Trade, { TradeType } from 'models/Trade'
import Order from 'models/Order'

const { AddressZero: ETHER_ADDRESS, Zero } = constants
export const TRADES_LIMIT = 20
export const ORDERS_LIMIT = 20

export interface ExchangeHandler {
  exchange: ExchangeType
  token: Token
  tradeListeners: (providers.Listener | null)[]
  createOrderListeners: (providers.Listener | null)[]
  cancelOrderListeners: (providers.Listener | null)[]
}

export const initializeContract = (provider: Web3Provider, token: Token): ExchangeHandler => {
  const exchangeAddress = process.env.REACT_APP_EXCHANGE_CONTRACT_ADDRESS
  if (!exchangeAddress) {
    throw 'Exchange address missing'
  }

  const exchange = new Contract(exchangeAddress, Exchange.abi, provider.getSigner()) as ExchangeType

  return {
    exchange,
    token,
    tradeListeners: [],
    createOrderListeners: [],
    cancelOrderListeners: [],
  }
}

export const getEthBalance = async ({ exchange }: ExchangeHandler, account: string): Promise<BigNumber> => {
  return (await exchange.ethBalanceOf(account)) as BigNumber
}

export const getTokenBalance = async ({ exchange, token }: ExchangeHandler, account: string): Promise<BigNumber> => {
  return (await exchange.tokenBalanceOf(token.address, account)) as BigNumber
}

export const subscribeTrades = (
  { exchange, tradeListeners }: ExchangeHandler,
  callback: (trade: Trade) => void
): number => {
  const listener = (
    orderId: BigNumber,
    _sellAccount: string,
    _sellToken: string,
    sellAmount: BigNumber,
    _buyAccount: string,
    buyToken: string,
    buyAmount: BigNumber,
    timestamp: BigNumber
  ) => {
    callback(getTradeFromTradeEvent({ orderId, sellAmount, buyToken, buyAmount, timestamp }))
  }

  exchange.on('Trade', listener)
  tradeListeners.push(listener)
  return tradeListeners.indexOf(listener)
}

export const unsubscribeTrades = ({ exchange, tradeListeners }: ExchangeHandler, id: number): void => {
  const listener = tradeListeners[id]
  if (!listener) {
    return
  }

  exchange.off('Trade', listener)
  tradeListeners[id] = null
}

export const subscribeCreateOrders = (
  { exchange, createOrderListeners }: ExchangeHandler,
  callback: (order: Order) => void
): number => {
  const listener = (
    id: BigNumber,
    account: string,
    _sellToken: string,
    sellAmount: BigNumber,
    buyToken: string,
    buyAmount: BigNumber,
    timestamp: BigNumber
  ) => {
    callback(getOrderFromCreateOrderEvent({ id, account, sellAmount, buyToken, buyAmount, timestamp }))
  }

  exchange.on('CreateOrder', listener)
  createOrderListeners.push(listener)
  return createOrderListeners.indexOf(listener)
}

export const unsubscribeCreateOrders = ({ exchange, createOrderListeners }: ExchangeHandler, id: number): void => {
  const listener = createOrderListeners[id]
  if (!listener) {
    return
  }

  exchange.off('CreateOrder', listener)
  createOrderListeners[id] = null
}

export const subscribeCancelOrders = (
  { exchange, cancelOrderListeners }: ExchangeHandler,
  callback: (orderId: BigNumber) => void
): number => {
  const listener = (id: BigNumber) => {
    callback(id)
  }

  exchange.on('CancelOrder', listener)
  cancelOrderListeners.push(listener)
  return cancelOrderListeners.indexOf(listener)
}

export const unsubscribeCancelOrders = ({ exchange, cancelOrderListeners }: ExchangeHandler, id: number): void => {
  const listener = cancelOrderListeners[id]
  if (!listener) {
    return
  }

  exchange.off('CancelOrder', listener)
  cancelOrderListeners[id] = null
}

export const getAllTrades = async ({ exchange }: ExchangeHandler): Promise<Trade[]> => {
  const events = await exchange.queryFilter(exchange.filters.Trade())
  return orderBy(
    events.slice(-TRADES_LIMIT).map(({ args }) => getTradeFromTradeEvent(args)),
    ['timestamp'],
    ['desc']
  )
}

export const getPendingOrders = async ({ exchange }: ExchangeHandler): Promise<Order[]> => {
  const events = await exchange.queryFilter(exchange.filters.CreateOrder())
  const cancelledIds = (await exchange.queryFilter(exchange.filters.CancelOrder())).map(({ args }) => args.id)
  const filledIds = (await exchange.queryFilter(exchange.filters.Trade())).map(({ args }) => args.orderId)
  const orders: Order[] = []
  for (let index = events.length - 1; index >= 0; index--) {
    const { args } = events[index]
    if (!cancelledIds.find((id) => id.eq(args.id)) && !filledIds.find((id) => id.eq(args.id))) {
      orders.push(getOrderFromCreateOrderEvent(args))
      if (orders.length >= ORDERS_LIMIT) {
        break
      }
    }
  }

  return orders
}

export const getTokenAllowance = async ({ token, exchange }: ExchangeHandler, account: string): Promise<BigNumber> => {
  return await token.allowance(account, exchange.address)
}

export const approveToken = async ({ token, exchange }: ExchangeHandler, amount: BigNumber): Promise<void> => {
  await token.approve(exchange.address, amount)
}

export const depositEther = async ({ exchange }: ExchangeHandler, amount: BigNumber): Promise<void> => {
  await exchange.depositEther({ value: amount })
}

export const depositToken = async ({ exchange, token }: ExchangeHandler, amount: BigNumber): Promise<void> => {
  await exchange.depositToken(token.address, amount)
}

export const withdrawEther = async ({ exchange }: ExchangeHandler, amount: BigNumber): Promise<void> => {
  await exchange.withdrawEther(amount)
}

export const withdrawToken = async ({ token, exchange }: ExchangeHandler, amount: BigNumber): Promise<void> => {
  await exchange.withdrawToken(token.address, amount)
}

const getTradeFromTradeEvent = ({
  orderId,
  sellAmount,
  buyToken,
  buyAmount,
  timestamp,
}: {
  orderId: BigNumber
  sellAmount: BigNumber
  buyToken: string
  buyAmount: BigNumber
  timestamp: BigNumber
}): Trade => {
  const type = getTradeType(buyToken)
  const totalPrice = type == TradeType.Buy ? sellAmount : buyAmount // 0.001
  const tokenAmount = type == TradeType.Buy ? buyAmount : sellAmount // 0.1
  const unitPrice = tokenAmount.isZero() ? Zero : totalPrice.mul(BigNumber.from(10).pow(DECIMALS)).div(tokenAmount)

  return {
    orderId: orderId.toString(),
    timestamp: getDateFromUnixTimestamp(timestamp),
    type,
    amount: tokenAmount.toString(),
    unitPrice: unitPrice.toString(),
    totalPrice: totalPrice.toString(),
  }
}

const getOrderFromCreateOrderEvent = ({
  id,
  account,
  sellAmount,
  buyToken,
  buyAmount,
  timestamp,
}: {
  id: BigNumber
  account: string
  sellAmount: BigNumber
  buyToken: string
  buyAmount: BigNumber
  timestamp: BigNumber
}): Order => {
  const type = getTradeType(buyToken)
  const totalPrice = type == TradeType.Buy ? sellAmount : buyAmount // 0.001
  const tokenAmount = type == TradeType.Buy ? buyAmount : sellAmount // 0.1
  const unitPrice = tokenAmount.isZero() ? Zero : totalPrice.mul(BigNumber.from(10).pow(DECIMALS)).div(tokenAmount)

  return {
    orderId: id.toString(),
    timestamp: getDateFromUnixTimestamp(timestamp),
    type,
    account,
    amount: tokenAmount.toString(),
    unitPrice: unitPrice.toString(),
    totalPrice: totalPrice.toString(),
  }
}

const getTradeType = (buyToken: string) => {
  return buyToken == ETHER_ADDRESS ? TradeType.Sell : TradeType.Buy
}
