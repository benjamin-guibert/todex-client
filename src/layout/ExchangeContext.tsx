import { createContext, useCallback, useEffect, useRef, useState } from 'react'
import { Web3Provider } from '@ethersproject/providers'
import { Token } from 'contracts/Token'
import { ExchangeHandler, getEthBalance, getTokenBalance, initializeContract } from 'libraries/contracts/exchange'

export interface ExchangeContextValue {
  initialized: boolean
  ethBalance: string | undefined
  tokenBalance: string | undefined
  initialize: (provider: Web3Provider, token: Token) => Promise<boolean>
  setAccount: (account: string | undefined) => void
  updateBalances: () => Promise<void>
}

export const useExchangeContext = (): ExchangeContextValue => {
  const exchangeHandlerRef = useRef<ExchangeHandler>()
  const [initialized, setInitialized] = useState<boolean>(false)
  const [ethBalance, setEthBalance] = useState<string | undefined>()
  const [account, setAccount] = useState<string | undefined>()
  const [tokenBalance, setTokenBalance] = useState<string | undefined>()

  const initializeValue = async (provider: Web3Provider, token: Token) => {
    exchangeHandlerRef.current = initializeContract(provider, token)
    setInitialized(!!exchangeHandlerRef.current)

    return !!exchangeHandlerRef.current
  }

  const updateBalancesValue = useCallback(async () => {
    if (!exchangeHandlerRef.current || !account) {
      return
    }
    setEthBalance((await getEthBalance(exchangeHandlerRef.current, account))?.toString())
    setTokenBalance((await getTokenBalance(exchangeHandlerRef.current, account))?.toString())
  }, [account])

  const setAccountValue = (account: string | undefined) => {
    setAccount(account)
  }

  useEffect(() => {
    updateBalancesValue()
  }, [updateBalancesValue, account])

  return {
    initialized,
    ethBalance,
    tokenBalance,
    initialize: initializeValue,
    setAccount: setAccountValue,
    updateBalances: updateBalancesValue,
  }
}

export const ExchangeContext = createContext<ExchangeContextValue>({
  initialized: false,
  ethBalance: undefined,
  tokenBalance: undefined,
  initialize: async () => false,
  setAccount: () => undefined,
  updateBalances: async () => undefined,
})
