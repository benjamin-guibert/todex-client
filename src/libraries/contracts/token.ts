import { ethers } from 'ethers'
import { Web3Provider } from '@ethersproject/providers'
import Token from 'contracts/abis/Token.json'
import { Token as TokenType } from 'contracts/Token'

export interface TokenHandler {
  provider: Web3Provider
  token: TokenType
}

export const initializeContract = (provider: Web3Provider): TokenHandler => {
  const tokenAddress = process.env.REACT_APP_TOKEN_CONTRACT_ADDRESS
  if (!tokenAddress) {
    throw new Error('Token address missing')
  }

  const token = new ethers.Contract(tokenAddress, Token.abi, provider.getSigner()) as TokenType

  return {
    provider,
    token,
  }
}
