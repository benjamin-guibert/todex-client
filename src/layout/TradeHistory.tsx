import React, { FC, useContext } from 'react'
import moment from 'moment'
import { printAmount } from 'libraries/helpers'
import { ExchangeContext } from './ExchangeContext'
import Table from 'react-bootstrap/Table'
import StyledTradeHistory from './TradeHistory.style'

const TradeHistory: FC = () => {
  const { trades } = useContext(ExchangeContext)

  return (
    <StyledTradeHistory>
      <Table variant="dark">
        <thead>
          <tr className="text-center">
            <th>Time</th>
            <th>Amount</th>
            <th>Price</th>
            <th>Total</th>
          </tr>
        </thead>
        <tbody>
          {trades.map(({ orderId, amount, unitPrice, totalPrice, timestamp }) => {
            return (
              <tr key={orderId} className="font-monospace">
                <td className="text-center text-muted">{moment(timestamp).format('HH:mm')}</td>
                <td className="text-end">{printAmount(amount)}</td>
                <td className="text-end">{printAmount(unitPrice, { maxDecimals: 5 })}</td>
                <td className="text-end">{printAmount(totalPrice)}</td>
              </tr>
            )
          })}
        </tbody>
      </Table>
    </StyledTradeHistory>
  )
}

export default TradeHistory
