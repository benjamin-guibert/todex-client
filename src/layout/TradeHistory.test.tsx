import * as React from 'react'
import TradeHistory from './TradeHistory'
import renderer from 'react-test-renderer'
import { BigNumber } from 'ethers'
import { parseEther } from 'ethers/lib/utils'
import { TradeType } from '../libraries/contracts/exchange'
import { InitializationStatus, MetaMaskContext, MetaMaskContextValue } from './MetaMaskContext'
import { ExchangeContext, ExchangeContextValue } from './ExchangeContext'

const createMetaMaskContextValue = (): MetaMaskContextValue => {
  return {
    account: undefined,
    balance: undefined,
    initializationStatus: InitializationStatus.Initialized,
    initialize: jest.fn(),
    connect: jest.fn(),
    getAccount: jest.fn(),
    getBalance: jest.fn(),
  }
}

const createExchangeContextValue = (): ExchangeContextValue => {
  return {
    initialized: true,
    ethBalance: undefined,
    tokenBalance: undefined,
    trades: [],
    initialize: jest.fn(),
    setAccount: jest.fn(),
    updateBalances: jest.fn(),
  }
}

const renderComponent = (metaMaskContextValue: MetaMaskContextValue, exchangeContextValue: ExchangeContextValue) =>
  renderer.create(
    <MetaMaskContext.Provider value={metaMaskContextValue}>
      <ExchangeContext.Provider value={exchangeContextValue}>
        <TradeHistory />
      </ExchangeContext.Provider>
    </MetaMaskContext.Provider>
  )

describe('<TradeHistory />', () => {
  it('should render when trades', () => {
    const metaMaskContextValue = createMetaMaskContextValue()
    const exchangeContextValue = {
      ...createExchangeContextValue(),
      trades: [
        {
          orderId: '2',
          timestamp: new Date(BigNumber.from('1612345678').toNumber()),
          type: TradeType.Buy,
          amount: parseEther('1000'),
          unitPrice: parseEther('0.001').toString(),
          totalPrice: parseEther('1'),
        },
        {
          orderId: '1',
          timestamp: new Date(BigNumber.from('1612344678').toNumber()),
          type: TradeType.Sell,
          amount: parseEther('1000'),
          unitPrice: parseEther('0.001').toString(),
          totalPrice: parseEther('1'),
        },
      ],
    }
    const result = renderComponent(metaMaskContextValue, exchangeContextValue)

    expect(result.toJSON()).toMatchSnapshot()
  })

  it('should render when no trades', () => {
    const metaMaskContextValue = createMetaMaskContextValue()
    const exchangeContextValue = createExchangeContextValue()
    const result = renderComponent(metaMaskContextValue, exchangeContextValue)

    expect(result.toJSON()).toMatchSnapshot()
  })
})
