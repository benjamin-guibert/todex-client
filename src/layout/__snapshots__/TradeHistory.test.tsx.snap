// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TradeHistory /> should render when no trades 1`] = `
<div
  className="sc-bdvvtL iytLv container"
>
  <table
    className="table table-dark"
  >
    <thead>
      <tr
        className="text-center"
      >
        <th>
          Time
        </th>
        <th>
          Amount
        </th>
        <th>
          Price
        </th>
        <th>
          Total
        </th>
      </tr>
    </thead>
    <tbody />
  </table>
</div>
`;

exports[`<TradeHistory /> should render when trades 1`] = `
<div
  className="sc-bdvvtL iytLv container"
>
  <table
    className="table table-dark"
  >
    <thead>
      <tr
        className="text-center"
      >
        <th>
          Time
        </th>
        <th>
          Amount
        </th>
        <th>
          Price
        </th>
        <th>
          Total
        </th>
      </tr>
    </thead>
    <tbody>
      <tr
        className="font-monospace"
      >
        <td
          className="text-center text-muted"
        >
          15:52
        </td>
        <td
          className="text-end"
        >
          1,000
        </td>
        <td
          className="text-end"
        >
          0.001
        </td>
        <td
          className="text-end"
        >
          1
        </td>
      </tr>
      <tr
        className="font-monospace"
      >
        <td
          className="text-center text-muted"
        >
          15:52
        </td>
        <td
          className="text-end"
        >
          1,000
        </td>
        <td
          className="text-end"
        >
          0.001
        </td>
        <td
          className="text-end"
        >
          1
        </td>
      </tr>
    </tbody>
  </table>
</div>
`;
